src/analysis/processing/qgsalgorithmrandomextract.cpp

```diff
@@ -102,42 +101,74 @@ QVariantMap QgsRandomExtractAlgorithm::processAlgorithm( const QVariantMap &para
     number = static_cast< int >( std::ceil( number * count / 100 ) );
   }
 
+  // Build a list of all feature ids
+  QgsFeatureIterator fit = source->getFeatures( QgsFeatureRequest()
+                           .setFlags( QgsFeatureRequest::NoGeometry )
+                           .setNoAttributes() );
+  std::vector< QgsFeatureId > allFeats;
+  allFeats.reserve( count );
+  QgsFeature f;
+  feedback->pushInfo( QObject::tr( "Building list of all features..." ) );
+  while ( fit.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+      return QVariantMap();
+    allFeats.push_back( f.id() );
+  }
+  feedback->pushInfo( QObject::tr( "Done." ) );
+
   // initialize random engine
   std::random_device randomDevice;
-  const std::mt19937 mersenneTwister( randomDevice() );
-  const std::uniform_int_distribution<int> fidsDistribution( 0, count );
-
-  QVector< QgsFeatureId > fids( number );
-  std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
-
-  QHash< QgsFeatureId, int > idsCount;
-  for ( const QgsFeatureId id : fids )
+  std::mt19937 mersenneTwister( randomDevice() );
+  std::uniform_int_distribution<size_t> fidsDistribution;
+
+  // If the number of features to select is greater than half the total number of features
+  // we will instead randomly select features to *exclude* from the output layer
+  size_t shuffledFeatureCount = number;
+  bool invertSelection = number > count / 2;
+  if ( invertSelection )
+    shuffledFeatureCount = count - number;
```

Comment:
There's a chance that the featureCount retrieved earlier is only an estimate, so let's use the actual count here instead:

```suggestion
  size_t actualFeatureCount = allFeats.size();
  size_t shuffledFeatureCount = number;
  bool invertSelection = number > actualFeatureCount / 2;
  if ( invertSelection )
    shuffledFeatureCount = actualFeatureCount - number;
```